### ISSUE TITLE
More secure implementation for Islands Architecture

### ISSUE THREAD

**Issue by arkark** [denoland/fresh#2254]
Hello, I'd like to share the following security considerations (which I don't think are vulnerabilities of Fresh) to make Fresh more secure.

- Props Manipulation for Islands Architecture
- Prototype Pollution in deserialization

This issue is long and some information may be missing. If you have any questions, feel free to ask me.

### Props Manipulation for Islands Architecture

Fresh uses Islands Architecture, and the following article introduces how islands work in Fresh:

- [A Gentle Introduction to Islands - Deno](https://deno.com/blog/intro-to-islands)

An example of generated client-side scripts is as follows (formatted):
```html
<script type="module" nonce="...hex...">
  import { deserialize } from "/_frsh/js/...hex.../deserializer.js";
  import { signal } from "/_frsh/js/...hex.../signals.js";

  const ST = document.getElementById("__FRSH_STATE").textContent;
  const STATE = deserialize(ST, signal);

  import { revive } from "/_frsh/js/...hex.../main.js";
  /* ... snip (import statements for islands) ... */

  const propsArr = typeof STATE !== "undefined" ? STATE[0] : [];
  revive({/* ... snip (a object for islands) ... */}, propsArr);
</script>
```

Fresh renders island components with a JSON value of `document.getElementById("__FRSH_STATE").textContent`. So, if users can inject an HTML element with `id="__FRSH_STATE"`, it is possible to manipulate the rendering process and potentially change the behavior of the application.

For some applications, there are cases where such an assumption is valid. For example, consider an application that allows users to write markdown text freely.

The following documentation introduces how to render markdown and this example uses [deno-gfm](https://github.com/denoland/deno-gfm):

- https://fresh.deno.dev/docs/examples/rendering-markdown

deno-gfm prevents XSS attacks with [sanitize-html](https://github.com/apostrophecms/sanitize-html), but allows adding `id` attributes to HTML elements:

- ref. https://github.com/denoland/deno-gfm/blob/0.2.5/mod.ts#L214-L219

If the following island is defined,
```typescript
import type { Signal } from "@preact/signals";
import { render } from "$gfm";

interface PreviewProps {
  text: Signal<string>;
}

export default function Preview(props: PreviewProps) {
  return (
    <div
      class="markdown-body"
      dangerouslySetInnerHTML={{ __html: render(props.text.value) }}
    />
  );
}
```

the following `text` parameter is valid as a markdown text and it overwrites the island state:
```html
<h1 id="__FRSH_STATE">
{"v":{"0":[{"text":{"_f":"s","v":"some text"}}]}
</h1>
```

### Prototype Pollution in deserialization

Fresh deserializes a JSON value of `__FRSH_STATE` with:

- `deserialize`: https://github.com/denoland/fresh/blob/1.6.2/src/runtime/deserializer.ts#L21-L63

```typescript
export function deserialize(
  str: string,
  signal?: <T>(a: T) => Signal<T>,
): unknown {
  /* ...snip... */

  const { v, r } = JSON.parse(str, reviver);
  const references = (r ?? []) as [string[], ...string[][]][];
  for (const [targetPath, ...refPaths] of references) {
    const target = targetPath.reduce((o, k) => k === null ? o : o[k], v);
    for (const refPath of refPaths) {
      if (refPath.length === 0) throw new Error("Invalid reference");
      // set the reference to the target object
      const parent = refPath.slice(0, -1).reduce(
        (o, k) => k === null ? o : o[k],
        v,
      );
      parent[refPath[refPath.length - 1]!] = target;
    }
  }
  return v;
}
```

It is assumed that users cannot directly specify the `str` parameter of this function. However, this assumption is invalid if users can overwrite the JSON value as described above. Then it is vulnerable to Prototype Pollution attacks because the `deserialize` function has no check for the attacks.

### Suggestions

Similar security risks may exist in other web frameworks using Islands Architecture. However, some of the frameworks I checked seemed realistically difficult to attack for some reasons:

- [Astro](https://github.com/withastro/astro)
    - Astro defines custom elements `<astro-island>` and renders island components using them.
        - https://github.com/withastro/astro/blob/create-astro@4.3.0/packages/astro/src/runtime/server/astro-island.ts#L48-L49
    - There are very few cases where users can insert a **custom element** if the application prevents XSS attacks (e.g. HTML sanitization).
- [Qwik](https://github.com/BuilderIO/qwik)
    - Qwik parses JSON values of `<script type="qwik/json">{...json...}</script>` and renders island components using it.
        - https://qwik.builder.io/docs/advanced/qrl/#rendered-html
        - https://github.com/BuilderIO/qwik/blob/v1.2.14/packages/qwik/src/qwikloader.ts#L31-L45
    - There are very few cases where users can insert a **script element** with an arbitrary id attribute if the application prevents XSS attacks (e.g. HTML sanitization).

To make Fresh more secure, I suggest the two measures:

- Change the method to get a JSON of `__FRSH_STATE`,  from `document.getElementById("__FRSH_STATE").textContent`.
    - The above framework implementation may be helpful.
- Add prototype pollution checks to `deserialize`.
    - FYI: [Prototype Pollution Prevention - OWASP Cheat Sheet Series](https://cheatsheetseries.owasp.org/cheatsheets/Prototype_Pollution_Prevention_Cheat_Sheet.html)
 
### Notes

I created a CTF challenge [LemonMD](https://github.com/arkark/my-ctf-challenges/tree/main/challenges/202312_SECCON_CTF_2023_Finals/web/lemonmd) focused on the above considerations. There is a vulnerable application and attackers can perform XSS.

Reproduction steps for the XSS:

1. Download the source code:
    - https://github.com/arkark/my-ctf-challenges/tree/main/challenges/202312_SECCON_CTF_2023_Finals/web/lemonmd/build/web
2. Launch the web server:
    - `docker run -it --rm -p 3000:3000 (docker build -q .)`
3. Access the service (`http://localhost:3000`).
4. Render the following markdown text:
    ```html
    <h1 id="__FRSH_STATE">
    {
      "v": {
        "0": [{
          "text":{
            "_f": "s",
            "v": "&lt;img src=0 onerror=\"alert(1)\"&gt;"
          }
        }],
        "*": ["onerror"]
      },
      "r": [
        [["*"], ["constructor", "prototype", "*"]]
      ]
    }
    </h1>
    ```
5. The `alert(1)` will be executed by the XSS.

I wrote the description for the challenge and the steps in detail:

- https://blog.arkark.dev/2023/12/28/seccon-finals/#web-LemonMD

**Comment by lucacasonato** [comment]
Thanks for the report. We're looking into this.

**Comment by marvinhagemeister** [comment]
Thanks for reporting these issues and the detailed description. We resolved the points with #2256 and #2255 and published a new Fresh patch release.

**Comment by arkark** [comment]
Thanks for the quick fix and the release!
